fastseq/optimizer/transformers/modeling_gpt2_optimizer.py

```diff
@@ -72,17 +72,17 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
             attn_output = torch.matmul(attn_weights, value)
             return attn_output, attn_weights
 
-        split_weights = attn_weights.split(self.cache_input_len, dim=-1)
-        w1 = split_weights[0]
+        w1 = attn_weights[:, :, :, :self.cache_input_len]
+        w2 = attn_weights[:, :, :, self.cache_input_len:]
         w1 = w1.view(
             (w1.size(0)//self.num_beams, self.num_beams) + w1.shape[1:])
         attn = torch.einsum(
             "bmhtd,bnhds->bmhts",
             w1,
             self.cache_input_value)
         attn = attn.reshape((-1,) + attn.shape[2:])
-        if len(split_weights) == 2:
-            attn += torch.matmul(split_weights[1], value)
+        if w2.shape[-1] == value.shape[-2]: 
```

Comment:
Will it be better to use `assert` here? Because they are supposed to be same. Otherwise, it means there are some issues in the code?

Reply:
Is this true on the first iteration also? Before there is anything in the cache? My original understanding was that the if was intended to exclude the case where len(split_weights) == 1